BCDA-282: GORM and bulk token revocation

knollfear opened the pull request:
<!--

--- PR Hygiene Checklist ---

1. Make sure your branch is named with this format: `user-initials/description-ABC-123`. For example, `jj/add-awesomeness-bcda-99999`
2. Update the PR title: `bcda-99999 Feature: Add Awesomeness`
3. Edit the text below - do not leave placeholders in the text.
4. Add any other details that will be helpful for the reviewers: details description, screenshots, etc
5. Request a review from someone/multiple someones
-->

<!-- Replace xxx with the JIRA ticket number: -->
### Fixes [BCDA-282](https://jira.cms.gov/browse/BCDA-282)
### Fixes [BCDA-188](https://jira.cms.gov/browse/BCDA-188)
### Fixes [BCDA-189](https://jira.cms.gov/browse/BCDA-189)

<!-- describe the problem being solved here -->

Remove XO and add in GORM
Improve Unit Test coverage
Add method for invalidating all tokens for a given user
add method for invalidating all tokens for an ACO
### Proposed changes:
<!-- List of changes with bullet points here: -->
GORM
Remove XO
Invalidating User TOkens
Invalidating ACO tokens
### Change Details
<!-- add detailed discussion of changes here: -->


### Security Implications
<!-- Does the change deal with PII at all? What should reviewers look for in terms of security concerns? -->
Check out the tests, they handle token security which is important


### Acceptance Validation
<!-- were you able to fully test the acceptance criteria on the related ticket? if not, why not? -->

<!-- insert screenshots if applicable (drag images here) -->


### Feedback Requested
<!-- what type of feedback you want from your reviewers? -->
Pleas ask questions about the GORM process if it isn't clear.  It is a change from XO, but it is much simpler and easier to code.

msnook commented:
If we are deprecating `xo`, can we update this target in the Makefile to use `GORM`?  Maybe this process should be tested?  
https://github.com/CMSgov/bcda-app/blob/master/Makefile#L5

msnook commented:
The README references `xo`.  Can that be removed?

knollfear commented:
GORM is not a CLI program, but a dependency in the project.  It should not require anything except proper inclusion in the vendor folder on build.

Removed XO from readme.md
Removed go get from the docker files
Added initialize to main.go to get the DB set up correctly

msnook commented:
I was looking at line 5 of the Makefile. 
https://github.com/CMSgov/bcda-app/blob/master/Makefile#L5

Should this command line invocation of `xo` be replaced?

msnook commented on Makefile:
```diff
@@ -1,9 +1,3 @@
-	docker-compose up -d db
-	echo "Waiting for db to be ready..."
-	sleep 5
-	PGSSLMODE=disable xo postgresql://postgres:toor@localhost:5432/bcda -o models
```
Now that `xo` has been completely removed from Makefile, does the `models` target need to be replaced with something that is GORM-specific?  Will we need to continue to execute the model generation in the future, or is this now completely obsolete?

knollfear replied:
completely obsolete.  GORM doesn't write go code the way XO did.  It handles dynamic query execution at run time instead of prewriting it all.

rnagle commented on Dockerfiles/Dockerfile.bcda:
```diff
@@ -7,6 +7,7 @@ RUN go get github.com/tools/godep
 RUN go get github.com/pressly/fresh
 RUN go get -u github.com/derekparker/delve/cmd/dlv
 
+
```
A couple extraneous newlines in the Dockerfiles.

knollfear replied:
removed in next commit.

knollfear replied:
removed

rnagle commented on bcda/auth/backend.go:
```diff
@@ -261,22 +226,48 @@ func getPublicKey() *rsa.PublicKey {
+}
 
+// Save a token to the DB for a user
+func (backend *JWTAuthenticationBackend) CreateToken(user bcdagorm.User) (bcdagorm.Token, error) {
```
What's the plan for this function? Should this be called when running the `bcda create-token` command? I only see it invoked in test cases. I'm probably just missing something.

knollfear replied:
I don't know exactly when this will be run.  I feel like we are still waiting for approval on our auth-X schemes.

knollfear replied:
I didn't want to change the function of this command.  I am happy to change it, but I wasn't sure if it was in scope or not.

rnagle commented on bcda/auth/backend.go:
```diff
@@ -261,22 +226,48 @@ func getPublicKey() *rsa.PublicKey {
+		Value:  tokenString,
+		Active: true,
 	}
+	db.Create(&token)
```
NB for the whole team: if we do start using this function as part of the `bcda create-token` command, we will be storing each generated token in our db.

knollfear replied:
It doesn't yet.  It can very easily though.

knollfear replied:
I left this function functionally unchanged so it will NOT persist a token just yet.   It probably should soon though.

rnagle commented on bcdagorm/models.go:
```diff
@@ -0,0 +1,61 @@
+
+	// Migrate the schema
+	// Add your new models here
+	db.AutoMigrate(
```
I'm a little worried about running migrations automatically when the app starts. Are these always safe to run? What happens when there are multiple instances of the app started across many servers?

Thinking it'd be nice to have this wrapped up in a `bcda migrate` command or similar to have control over the timing of migrations. What do you think?

I'm also wondering if there's a way to have gorm output the SQL it plans to apply for migrations. Maybe a [`bcda sqlmigrate`](https://docs.djangoproject.com/en/2.1/ref/django-admin/#django-admin-sqlmigrate) command?

knollfear replied:
I think a sql migrate command makes sense.  I will change it.  It needs to be added to the jenkins deploy though.

knollfear replied:
Added this command.

rnagle commented on bcda/auth/backend_test.go:
```diff
@@ -33,43 +40,233 @@ func (s *BackendTestSuite) TestCreateACO() {
 
 func (s *BackendTestSuite) TestCreateUser() {
-	userUUID, err := s.authBackend.CreateUser("First Last", "firstlast@example.com", uuid.Parse("DBBD1CE1-AE24-435C-807D-ED45953077D3"))
+	name, email, sampleUUID := "First Last", "firstlast@exaple.com", "DBBD1CE1-AE24-435C-807D-ED45953077D3"
```
All these tests! 🍔

rnagle commented on bcda/auth/backend.go:
```diff
@@ -261,22 +226,48 @@ func getPublicKey() *rsa.PublicKey {
+	token := bcdagorm.Token{
+		UUID:   uuid.Parse(claims["id"].(string)),
+		UserID: user.UUID,
+		Value:  tokenString,
```
One last request for me: please make sure we're hashing token values anywhere we are inserting/updating them in the database. Check out: https://github.com/CMSgov/bcda-app/blob/master/bcda/auth/backend.go#L135

knollfear replied:
Resolved in latest commit.  token.Value is now hashed on save IF it is a valid JWT token.  If it isn't then I assume it is already hashed and don't want to double hash it.

embh commented on bcda/auth/backend.go:
```diff
@@ -261,22 +226,51 @@ func getPublicKey() *rsa.PublicKey {
+	return token.Claims.(jwt.MapClaims)
+}
+
+func (backend *JWTAuthenticationBackend) GetJWTToken(tokenString string) (*jwt.Token, error) {
```
Not a big deal at all, but can we call this `GetJWT` or `GetToken`, since the T in JWT = Token?

knollfear replied:
getToken isn't going to work because that is ambiguous about what kind of token, our token object in the db vs a JWT token.  I think getJWT is also unclear because that could just be the JWT package or some JWT context instead of an actual JWT token.  JWT is a clear part of our system and has other objects like claims (See line 229) that we use.  I think getJWToken might be fine, but it isn't jwt.oken, it is jwt.Token.

knollfear replied:
Committed a change to GetJWToken
